### PR TITLE
Ordinal variables

### PR DESCRIPTION
**Pull request by xulaus**

### CODE REVIEW

File: CHAID/__main__.py [hunk 1]
```diff
@@ -14,7 +15,14 @@ def main():
...
     parser.add_argument('dependent_variable', nargs=1)
-    parser.add_argument('independent_variables', nargs='+')
+
+    var = parser.add_argument_group('Independent Variable Specification')
+    var.add_argument('nominal_variables', nargs='*', help='The names of '
+                     'independent variables to use that have no intrinsic '
+                     'order to them')
+    var.add_argument('--ordinal-variables', type=str, nargs='*',
+                     help='The names of independent variables to use that '
+                     'have an intrinsic order but a finite amount of states')
```
Comment (Rambatino) [inline CHAID/__main__.py]: ooo very nice descriptions 👍

File: CHAID/__main__.py [hunk 2]
```diff
@@ -29,16 +37,20 @@ def main():
...
-    # data = data.rename(columns=data.loc[0]).iloc[1:]
+    if nspace.file[-4:] == '.csv':
+        data = pd.read_csv(nspace.file)
+    elif nspace.file[-4:] == '.sav':
+        raw_data = spss.SavReader(nspace.file, returnHeader=True)
+        raw_data_list = list(raw_data)
+        data = pd.DataFrame(raw_data_list)
+        data = data.rename(columns=data.loc[0]).iloc[1:]
+    else:
+        print('Uknown file type')
```
Comment (Rambatino) [inline CHAID/__main__.py]: *Unknown

File: CHAID/column.py [hunk 3]
```diff
@@ -123,15 +124,102 @@ def __getitem__(self, key):
...
+    def possible_groupings(self):
+        if self._possible_groups is None:
+            ranges = sorted(self._groupings.items())
+            candidates = zip(ranges[0:], ranges[1:])
+            self._possible_groups = [
+                (k1, k2) for (k1, minmax1), (k2, minmax2) in candidates
+                if minmax1[1] == minmax2[0]
+            ]
+            if self._nan in self._arr:
+                self._possible_groups += list(
```
Comment (Rambatino) [inline CHAID/column.py]: Why does this need to be a list wrapped around it?

File: CHAID/tree.py [hunk 4]
```diff
@@ -61,15 +61,24 @@ class Tree(object):
...
+                 variable_types=None):
         self.alpha_merge = alpha_merge
         self.max_depth = max_depth
         self.min_parent_node_size = min_parent_node_size
         self.min_child_node_size = min_child_node_size
         self.split_titles = split_titles or []
         self.vectorised_array = []
-        for ind in range(0, ndarr.shape[1]):
-            self.vectorised_array.append(NominalColumn(ndarr[:, ind]))
+        variable_types = variable_types or ['nominal'] * ndarr.shape[1]
```
Comment (Rambatino) [inline CHAID/tree.py]: default to nominal?

File: CHAID/tree.py [hunk 5]
```diff
@@ -165,29 +176,30 @@ def generate_best_split(self, ind, dep, wt=None):
...
-                groupings = list(index.possible_groupings())
-                size = len(groupings)
+                    for dep_v in all_dep:
+                        freq[col][dep_v] = wt[(ind_var.arr == col) * (dep.arr == dep_v)].sum()
 
+            while next(ind_var.possible_groupings(), None) is not None:
                 sub_data_columns = [('combinations', object), ('p', float), ('chi', float)]
-                sub_data = np.array([(None, 0, 1)]*size, dtype=sub_data_columns, order='F')
-                for j, comb in groupings:
+                choice = None
```
Comment (Rambatino) [inline CHAID/tree.py]: comma separated, you no like?

File: tests/test_ordinal_column.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+from numpy import nan
+from setup_tests import list_ordered_equal,list_unordered_equal, CHAID
+
+NAN = float('nan')
+
+class TestOrdinalDeepCopy(TestCase):
+    """ Test fixture class for deep copy method """
+    def setUp(self):
+        """ Setup for copy tests"""
+        # Use string so numpy array dtype is object and may store references
```
Comment (Rambatino) [inline tests/test_ordinal_column.py]: what string?

File: tests/test_ordinal_column.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
+"""
```
Comment (Rambatino) [inline tests/test_ordinal_column.py]: dtype Object tests?
What happens when strings are passed in?

File: tests/test_tree.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ def test_best_split_with_combination():
     assert list_ordered_equal(ndarr, orig_ndarr), 'Calling chaid should have no side affects for original numpy arrays'
     assert list_ordered_equal(arr, orig_arr), 'Calling chaid should have no side affects for original numpy arrays'
     assert split.column_id == 0, 'Identifies correct column to split on'
-    assert list_unordered_equal(split.split_map, [[1], [2, 3]]), 'Correctly identifies catagories'
+    assert list_unordered_equal(split.split_map, [[1], [2], [3]]), 'Correctly identifies catagories'
```
Comment (Rambatino) [inline tests/test_tree.py]: what's this?

File: tests/test_ordinal_column.py (hunk 6 above)
Comment (xulaus) [inline tests/test_ordinal_column.py]: Copy paste error :X

File: tests/test_ordinal_column.py (hunk 7 above)
Comment (xulaus) [inline tests/test_ordinal_column.py]: Strings aren't supported as they have no strict ordering. Alphabetical is too weak an assumption to make.

Comment (Rambatino) [inline tests/test_ordinal_column.py]: yeah but what about numbers stored as objects?

File: CHAID/tree.py [hunk 9]
```diff
@@ -55,7 +55,8 @@ class Tree(object):
         the threshold value for the maximum number of levels after the root
         node in the tree (default 2)
     min_parent_node_size : float
-        the threshold value of the number of respondents that the node must
+        the threshold value of the number of respondents tha
```
Comment (Rambatino) [inline CHAID/tree.py]: newb

File: tests/test_ordinal_column.py (hunk 7 above)
Comment (xulaus) [inline tests/test_ordinal_column.py]: I do some jiggery pokery with casting to force them to `int`s. You are right in that I need tests for that though

File: CHAID/tree.py (hunk 9 above)
Comment (xulaus) [inline CHAID/tree.py]: Me no program good

**Comment by Rambatino** [comment]
Hmm it appears we have a numpy sorting issue: `ar.sort()`

``` python 
        if optional_indices:
            perm = ar.argsort(kind='mergesort' if return_index else 'quicksort')
            aux = ar[perm]
        else:
>           ar.sort()
E           TypeError: unorderable types: NoneType() > float()

/usr/local/lib/python3.5/site-packages/numpy/lib/arraysetops.py:198: TypeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /usr/local/lib/python3.5/site-packages/numpy/lib/arraysetops.py(198)unique()
-> ar.sort()
(Pdb) ar.sort()
*** TypeError: unorderable types: NoneType() > float()
(Pdb) ar
array([1.0, 2.0, 3.0, 3.0, 3.0, 3.0, 4.0, 5.0, 10.0, None], dtype=object)
```

**Comment by Rambatino** [comment]
@xulaus argh dealing with this again in python 3. So weird when I went on the numpy issues, found this issue and then my CHAID was linked in it! numpy/numpy#641

File: CHAID/column.py [hunk 10]
```diff
@@ -123,15 +123,106 @@ def __getitem__(self, key):
...
+            self._arr, self.orig_type = self.substitute_values(self._arr)
+
+        self._groupings = {}
+        if groupings is None:
+            for x in np.unique(self._arr):
+                self._groupings[x] = [x, x + 1, False]
+        else:
+            for x in np.unique(self._arr):
+                self._groupings[x] = list(groupings[x])
+        self._nan = np.array([np.nan]).astype(int)[0]
```
Comment (Rambatino) [inline CHAID/column.py:158]: @xulaus can you remember what this line is doing?

Comment (xulaus) [reply CHAID/column.py:158]: Not exactly. I think we were using nan as a sentinel value that can combine with any item in the ordinal but we wanted the arrays to all be of the same dtype so we did this force cast.

**Pull request merged**